### ISSUE TITLE
Feature: add other fuzzing platforms to the Fuzzing check

### ISSUE THREAD

**Issue by laurentsimon** [ossf/scorecard#897]
see title

**Comment by oliverchang** [comment]
Another one to add would be clusterfuzzlite which is launching soon: https://github.com/google/clusterfuzzlite.

**Comment by naveensrinivasan** [comment]
On a different note should we add this check to scorecard?

**Comment by oliverchang** [comment]
> Thanks Oliver. To detect clusterfuzzlite, we need to check for a workflow which uses `google/clusterfuzzlite/actions/run_fuzzers` action, is this correct? Do you expect users to enable it on PRs or push to main or something else?

Yep! Just check for the `google/clusterfuzzlite/actions/run_fuzzers` action. I think we can be flexible about where it's enabled (PR or push or otherwise) so we don't need to check that. 

> What about results of this check from this Fuzz?
> On a different note should we add this check to scorecard?

Could you clarify what you mean here? Are you suggesting a more detailed check beyond just checking if Fuzzing is enabled?

**Comment by naveensrinivasan** [comment]
> > Thanks Oliver. To detect clusterfuzzlite, we need to check for a workflow which uses `google/clusterfuzzlite/actions/run_fuzzers` action, is this correct? Do you expect users to enable it on PRs or push to main or something else?
> 
> Yep! Just check for the `google/clusterfuzzlite/actions/run_fuzzers` action. I think we can be flexible about where it's enabled (PR or push or otherwise) so we don't need to check that.
> 
> > What about results of this check from this Fuzz?
> > On a different note should we add this check to scorecard?
> 
> @naveensrinivasan Could you clarify what you mean here? Are you suggesting a more detailed check beyond just checking if Fuzzing is enabled?

Yes , how do we report the results of these fuzz runs?

**Comment by laurentsimon** [comment]
> > Thanks Oliver. To detect clusterfuzzlite, we need to check for a workflow which uses `google/clusterfuzzlite/actions/run_fuzzers` action, is this correct? Do you expect users to enable it on PRs or push to main or something else?
> 
> Yep! Just check for the `google/clusterfuzzlite/actions/run_fuzzers` action. I think we can be flexible about where it's enabled (PR or push or otherwise) so we don't need to check that.
> 
 great! Do you have cycles to add it before you officially release clusterfuzzlite? Or shall we add this to v4 milestone (~EOY)?

**Comment by laurentsimon** [comment]
@naveensrinivasan is there a special command/keyword we can use to say "this issue should not be closed automatically by the bot"?

**Comment by georgettica** [comment]
I wanted to also not that `go1.18` has fuzzing out of the box. didn't know if it should be included in this issue or a different one

**Comment by laurentsimon** [comment]
I think it fits in this issue. If you know how to check for its use (there's a command for it?) please let us know. Feel free to send. aPR if you have time for it. Thanks!

**Comment by georgettica** [comment]
I think oss-fuzz integrates with the native fizzing, but not sure how exactly.
Worth checking unrelated to my investigation

**Comment by laurentsimon** [comment]
We need to look for the `--fuzz` command in the workflow https://go.dev/doc/fuzz/. The assumption is that the command is used directly in the workflow, and not in a script that's invoked via the workflow. We can start simple and then iterate, though

**Comment by azeemshaikh38** [comment]
@georgettica @laurentsimon could we start with checking for `*_test.go` files in the repository which contain functions matching the regex `func Fuzz* (* \*testing.F)` ? Would that be a good start for adding this support?

**Comment by georgettica** [comment]
I think `go test -fuzz` takes any string and it's convention to use the prefix Fuzz
Not that I am opposed to the ideas, but letting you know it's based on convention rather than validation

**Comment by azeemshaikh38** [comment]
https://go.dev/doc/fuzz/ mentions this as a requirement. But maybe this is not a hard requirement?

![image](https://user-images.githubusercontent.com/4825078/161808658-1f1bf4b8-39bc-488b-a5df-3bbfde337035.png)

**Comment by georgettica** [comment]
yeah, same (verified aswell)
expected the `og test -fuzz=Bar` to at least. warn me

**Comment by laurentsimon** [comment]
I think it's fair to start with the regex @azeemshaikh38 provided. The warning is something Go team should fix, rather than us, no? Note that a while back, we were contemplating having a `Unit-Test` checks (or part of an existing check) so the work you're doing could be useful in this context too.

**Comment by georgettica** [comment]
Yup, it's a go team thing. Noting it here as I didn't start the conversation there

**Comment by oliverchang** [comment]
If we are extending the Fuzzing check to include looking for fuzzer definitions for Go (which I think is a good idea!), we should also extend this to other languages, which all have a consistent way to do this that we can detect. 

e.g. for C/C++, we can look for instances of LLVMFuzzerTestOneInput in c/cc/cxx/cpp files.

**Comment by inferno-chromium** [comment]
@oliverchang - can someone from fuzzing team pick this, it will be nice to have this scorecard check comprehensive.

**Comment by oliverchang** [comment]
@navidem -- this (https://github.com/ossf/scorecard/issues/897#issuecomment-1089849918) might be a interesting side project to tackle!

**Comment by laurentsimon** [comment]
@Navidem feel free to ask questions if you need some pointers

**Comment by oliverchang** [comment]
Fuzzing patterns to add:

- C/C++: https://llvm.org/docs/LibFuzzer.html#fuzz-target
- Other languages: See the language subguides on the canonical ways to write fuzzers in other languages on https://google.github.io/oss-fuzz/getting-started/new-project-guide/

**Comment by aidenwang9867** [comment]
> Thanks @oliverchang !
> 
> /cc @aidenwang9867

Thanks Laurent & Oliver, I'll take the Go side. @laurentsimon @oliverchang

**Comment by aidenwang9867** [comment]
@oliverchang @Navidem

In the merged PR[#1979](https://github.com/ossf/scorecard/pull/1979), I added support for checking Go built-in fuzzers as a part of scorecard's fuzzing check. Most of the important changes have been made in `checks/raw/fuzzing.go`, please refer to the changes in this PR to see them.

Built-in fuzzer patterns for other programming languages can be added [here](https://github.com/ossf/scorecard/blob/main/checks/raw/fuzzing.go#L50). I'll work on another feature for a while but I might implement some fuzzing supports for other languages if I have time. I'll ping you here if I start to do that. :)

cc @laurentsimon

**Comment by balteravishay** [comment]
what about [One-Fuzz](https://github.com/microsoft/onefuzz)? Any chance to support that by Scorecard?

**Comment by laurentsimon** [comment]
Absolutely. What does the integration in a repository look like? 
For example, this is how we check for clusterfuzzlite: https://github.com/ossf/scorecard/blob/69eb1ccf1d0cf8c5b291044479f18672bf250325/checks/raw/fuzzing.go#L120

If you tell us what we should be looking for in thee repo, we can add a function to check for it in the file linked above.

**Comment by balteravishay** [comment]
@laurentsimon, sorry for the delay, how about something like that: https://github.com/ossf/scorecard/pull/2141 :)

**Comment by DavidKorczynski** [comment]
Added support for C, CPP, Python, Java and Rust fuzzing in https://github.com/ossf/scorecard/pull/3473

Also opened up a discussion for splitting fuzzing checks in two categories: (1) a check for the presence of source code for a fuzzer and (2) a check for whether a project is continuously fuzzed. Issue is here: https://github.com/ossf/scorecard/issues/3475
